<!DOCTYPE html>
<html>
<head>
<title>README.md</title>
<meta http-equiv="Content-type" content="text/html;charset=UTF-8">

<style>
/* https://github.com/microsoft/vscode/blob/master/extensions/markdown-language-features/media/markdown.css */
/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license information.
 *--------------------------------------------------------------------------------------------*/

body {
	font-family: var(--vscode-markdown-font-family, -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif);
	font-size: var(--vscode-markdown-font-size, 14px);
	padding: 0 26px;
	line-height: var(--vscode-markdown-line-height, 22px);
	word-wrap: break-word;
}

#code-csp-warning {
	position: fixed;
	top: 0;
	right: 0;
	color: white;
	margin: 16px;
	text-align: center;
	font-size: 12px;
	font-family: sans-serif;
	background-color:#444444;
	cursor: pointer;
	padding: 6px;
	box-shadow: 1px 1px 1px rgba(0,0,0,.25);
}

#code-csp-warning:hover {
	text-decoration: none;
	background-color:#007acc;
	box-shadow: 2px 2px 2px rgba(0,0,0,.25);
}

body.scrollBeyondLastLine {
	margin-bottom: calc(100vh - 22px);
}

body.showEditorSelection .code-line {
	position: relative;
}

body.showEditorSelection .code-active-line:before,
body.showEditorSelection .code-line:hover:before {
	content: "";
	display: block;
	position: absolute;
	top: 0;
	left: -12px;
	height: 100%;
}

body.showEditorSelection li.code-active-line:before,
body.showEditorSelection li.code-line:hover:before {
	left: -30px;
}

.vscode-light.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(0, 0, 0, 0.15);
}

.vscode-light.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(0, 0, 0, 0.40);
}

.vscode-light.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-dark.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 255, 255, 0.4);
}

.vscode-dark.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 255, 255, 0.60);
}

.vscode-dark.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

.vscode-high-contrast.showEditorSelection .code-active-line:before {
	border-left: 3px solid rgba(255, 160, 0, 0.7);
}

.vscode-high-contrast.showEditorSelection .code-line:hover:before {
	border-left: 3px solid rgba(255, 160, 0, 1);
}

.vscode-high-contrast.showEditorSelection .code-line .code-line:hover:before {
	border-left: none;
}

img {
	max-width: 100%;
	max-height: 100%;
}

a {
	text-decoration: none;
}

a:hover {
	text-decoration: underline;
}

a:focus,
input:focus,
select:focus,
textarea:focus {
	outline: 1px solid -webkit-focus-ring-color;
	outline-offset: -1px;
}

hr {
	border: 0;
	height: 2px;
	border-bottom: 2px solid;
}

h1 {
	padding-bottom: 0.3em;
	line-height: 1.2;
	border-bottom-width: 1px;
	border-bottom-style: solid;
}

h1, h2, h3 {
	font-weight: normal;
}

table {
	border-collapse: collapse;
}

table > thead > tr > th {
	text-align: left;
	border-bottom: 1px solid;
}

table > thead > tr > th,
table > thead > tr > td,
table > tbody > tr > th,
table > tbody > tr > td {
	padding: 5px 10px;
}

table > tbody > tr + tr > td {
	border-top: 1px solid;
}

blockquote {
	margin: 0 7px 0 5px;
	padding: 0 16px 0 10px;
	border-left-width: 5px;
	border-left-style: solid;
}

code {
	font-family: Menlo, Monaco, Consolas, "Droid Sans Mono", "Courier New", monospace, "Droid Sans Fallback";
	font-size: 1em;
	line-height: 1.357em;
}

body.wordWrap pre {
	white-space: pre-wrap;
}

pre:not(.hljs),
pre.hljs code > div {
	padding: 16px;
	border-radius: 3px;
	overflow: auto;
}

pre code {
	color: var(--vscode-editor-foreground);
	tab-size: 4;
}

/** Theming */

.vscode-light pre {
	background-color: rgba(220, 220, 220, 0.4);
}

.vscode-dark pre {
	background-color: rgba(10, 10, 10, 0.4);
}

.vscode-high-contrast pre {
	background-color: rgb(0, 0, 0);
}

.vscode-high-contrast h1 {
	border-color: rgb(0, 0, 0);
}

.vscode-light table > thead > tr > th {
	border-color: rgba(0, 0, 0, 0.69);
}

.vscode-dark table > thead > tr > th {
	border-color: rgba(255, 255, 255, 0.69);
}

.vscode-light h1,
.vscode-light hr,
.vscode-light table > tbody > tr + tr > td {
	border-color: rgba(0, 0, 0, 0.18);
}

.vscode-dark h1,
.vscode-dark hr,
.vscode-dark table > tbody > tr + tr > td {
	border-color: rgba(255, 255, 255, 0.18);
}

</style>

<style>
/* Tomorrow Theme */
/* http://jmblog.github.com/color-themes-for-google-code-highlightjs */
/* Original theme - https://github.com/chriskempson/tomorrow-theme */

/* Tomorrow Comment */
.hljs-comment,
.hljs-quote {
	color: #8e908c;
}

/* Tomorrow Red */
.hljs-variable,
.hljs-template-variable,
.hljs-tag,
.hljs-name,
.hljs-selector-id,
.hljs-selector-class,
.hljs-regexp,
.hljs-deletion {
	color: #c82829;
}

/* Tomorrow Orange */
.hljs-number,
.hljs-built_in,
.hljs-builtin-name,
.hljs-literal,
.hljs-type,
.hljs-params,
.hljs-meta,
.hljs-link {
	color: #f5871f;
}

/* Tomorrow Yellow */
.hljs-attribute {
	color: #eab700;
}

/* Tomorrow Green */
.hljs-string,
.hljs-symbol,
.hljs-bullet,
.hljs-addition {
	color: #718c00;
}

/* Tomorrow Blue */
.hljs-title,
.hljs-section {
	color: #4271ae;
}

/* Tomorrow Purple */
.hljs-keyword,
.hljs-selector-tag {
	color: #8959a8;
}

.hljs {
	display: block;
	overflow-x: auto;
	color: #4d4d4c;
	padding: 0.5em;
}

.hljs-emphasis {
	font-style: italic;
}

.hljs-strong {
	font-weight: bold;
}
</style>

<style>
/*
 * Markdown PDF CSS
 */

 body {
	font-family: -apple-system, BlinkMacSystemFont, "Segoe WPC", "Segoe UI", "Ubuntu", "Droid Sans", sans-serif, "Meiryo";
	padding: 0 12px;
}

pre {
	background-color: #f8f8f8;
	border: 1px solid #cccccc;
	border-radius: 3px;
	overflow-x: auto;
	white-space: pre-wrap;
	overflow-wrap: break-word;
}

pre:not(.hljs) {
	padding: 23px;
	line-height: 19px;
}

blockquote {
	background: rgba(127, 127, 127, 0.1);
	border-color: rgba(0, 122, 204, 0.5);
}

.emoji {
	height: 1.4em;
}

code {
	font-size: 14px;
	line-height: 19px;
}

/* for inline code */
:not(pre):not(.hljs) > code {
	color: #C9AE75; /* Change the old color so it seems less like an error */
	font-size: inherit;
}

/* Page Break : use <div class="page"/> to insert page break
-------------------------------------------------------- */
.page {
	page-break-after: always;
}

</style>

<script src="https://unpkg.com/mermaid/dist/mermaid.min.js"></script>
</head>
<body>
  <script>
    mermaid.initialize({
      startOnLoad: true,
      theme: document.body.classList.contains('vscode-dark') || document.body.classList.contains('vscode-high-contrast')
          ? 'dark'
          : 'default'
    });
  </script>
<h1 id="biosynthnexus">BioSynthNexus</h1>
<hr>
<h2 id="overview">Overview</h2>
<p>A graphical user interface to easily access and output protein data.</p>
<p>Extracts data from a
<strong>Genome Neighborhood Network (GNN)</strong> SQLITE file
generated by the Genome neighborhood tool (GNT) from
Enzyme Function Initiative (EFI) Tools.
For more information on generating the GNN see the
<a href="https://efi.igb.illinois.edu/">EFI website</a>.<sup>1,2</sup></p>
<p>Extracts <a href="https://www.uniprot.org/">UniProt</a>
data for a given protein Accession ID (or from a list of IDs).<sup>3</sup></p>
<hr>
<h2 id="running-installation-packaging">Running, Installation, Packaging</h2>
<h3 id="running-the-program-directly-from-an-executable">Running the Program Directly from an Executable</h3>
<p>Provided in the <code>dist</code> directory are pre-packaged executables that can be download for the user's operating system of choice.</p>
<h3 id="manual-installation--packaging">Manual Installation / Packaging</h3>
<p>An enviorment can be created with pipenv or conda using the provided Pipfile or requirements.txt</p>
<h4 id="pipenv">Pipenv</h4>
<ul>
<li>Open terminal in directory containing the repository</li>
<li>Run <code>pipenv install</code><br/>
<ul>
<li>All required dependencies should then be installed<br/></li>
</ul>
</li>
<li>The program can then be run with <code>pipenv run python main.py</code></li>
</ul>
<h4 id="conda">Conda</h4>
<ul>
<li>Open terminal in directory containing the repository</li>
<li>Create a conda enviorment <code>conda create -n env_name python=3.10 pip</code></li>
<li>Activate enviorment <code>conda activate env_name</code></li>
<li>Install Requirements <code>pip install -r requirements.txt</code></li>
<li>Run the program <code>python main.py</code></li>
</ul>
<h4 id="packaging-into-an-application-optional">Packaging into an application (optional)</h4>
<p>If you would like to package your modified code into a single executable, pyinstaller is included in the dev-packages of the Pipfile</p>
<ul>
<li>Open terminal in repositiory directory</li>
<li>Run <code>pipenv install -d</code>
<ul>
<li>Note: If you are using a conda enviorment, you have to manually install pyinstaller <code>pip install pyinstaller</code></li>
</ul>
</li>
<li>Run <code>pyinstaller --windowed --onefile --add-data=ui_main_window.ui:./ --add-data=custom_ui_theme.xml:./ main.py</code>
<ul>
<li>A Folder named <code>dist</code> will contain the packaged application</li>
</ul>
</li>
</ul>
<hr>
<h3 id="terminology">Terminology:</h3>
<p>Query Accessions - The accessions used to generate the SQLite File.
<br> Note: The query accession is the same as the BGC ID, but will be referred to individually throughout.</p>
<p>GNN - Genome Neighborhood Network</p>
<hr>
<h2 id="input--output-options">Input / Output Options:</h2>
<h3 id="uniprot-requests">UniProt Requests:</h3>
<table>
<thead>
<tr>
<th>Output</th>
<th>Input</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>FASTA</td>
<td>Accession ID(s)</td>
<td>Gets the FASTA Sequence(s)</td>
</tr>
<tr>
<td>GenBankID</td>
<td>Accession ID(s)</td>
<td>Gets the GenBank Nucleotide Sequence ID(s)</td>
</tr>
<tr>
<td>GenBank_Protein_ID</td>
<td>Accession ID(s)</td>
<td>Gets the GenBank Protein ID(s)</td>
</tr>
<tr>
<td>GenBank_ORF_ID</td>
<td>Accession ID(s)</td>
<td>Gets the GenBank Open Reading Frame (ORF) ID(s)</td>
</tr>
</tbody>
</table>
<h3 id="bgc-requests">BGC Requests:</h3>
<table>
<thead>
<tr>
<th>Output</th>
<th>Input</th>
<th>Description</th>
</tr>
</thead>
<tbody>
<tr>
<td>Query Accession</td>
<td>PFam ID(s)</td>
<td>Gets the Query Accession IDs that correspond to their BGCs containing the given PFam ID(s)</td>
</tr>
<tr>
<td>BGC ID</td>
<td>PFam ID(s)</td>
<td>Gets the BGC ID(s) for those that contain the given PFam ID(s)</td>
</tr>
<tr>
<td>BGC-PFam</td>
<td>BGC ID</td>
<td>Gets the PFams for each of the proteins within a single BGC</td>
</tr>
<tr>
<td>BGC-Accession</td>
<td>BGC ID</td>
<td>Gets the Accession IDs for each of the proteins within a single BGC</td>
</tr>
<tr>
<td>Accessions from BGC by PFam</td>
<td>BGC ID(s)<br/> + PFam (Secondary Input)</td>
<td>Gets the Accession IDs for the protein within each BGC with the selected PFam </br> Displays as 'Output Accession_(BGC ID)'</td>
</tr>
<tr>
<td>BGC-PFam Similarity</td>
<td>PFam ID(s)<br/> Optional: PFam in Secondary Input</td>
<td>Searches all BGCs for each given PFam ID, Displays as 'BGC_ID(number of PFams it contains from the given input list)'<br/>The <code>Secondary Input</code> can be used to limit the searched BGCs to those that contain this inputed secondary PFam ID</td>
</tr>
</tbody>
</table>
<p>Notes:</p>
<ol>
<li>The <code>Accessions from BGC by PFam</code> outputs as <code>Output Accession_(BGC ID)</code>,
however if you use <code>Replace Input with Output</code>, only the Output Accession will be displayed.</li>
<li>If the input field is empty, BGCs that contain unannotated proteins will be considered.</li>
</ol>
<hr>
<h2 id="bgc-request">BGC Request:</h2>
<ul>
<li>Click the <code>Upload</code> button to select your GNN *.sqlite file</li>
<li>Change the Request Type to <code>BGC</code></li>
<li>Select your desired output (See <strong>BGC Requests</strong> above for guidance)</li>
<li>Fill the left text field, labelled 'Input', with respective input</li>
<li>Click <code>Search</code></li>
<li>Output will be displayed in the right text field, labelled 'Output'</li>
<li>The text in the Output Box can be used as an Input by utilizing the <code>Replace Input with Output</code> Button</li>
</ul>
<hr>
<h2 id="usage-examples">Usage examples:</h2>
<h3 id="example-1">Example 1:</h3>
<p>You have a GNN of the query accessions, you want to search for all BGCs which contain proteins assigned with the PFam <code>PF...</code>.
Then you would like to get the accession IDs associated with the Pfam proteins that you just found.
Lastly, you would like the FASTA sequence for each of those proteins to compare.</p>
<h4 id="get-the-bgcs-that-contain-the-pfams-of-interest">Get the BGCs that contain the PFam(s) of interest</h4>
<ol>
<li>Upload the GNN generated for protein X</li>
<li>Enter the PFam(s) into the input field</li>
<li>Select <code>BGC ID</code> as the Output Type</li>
<li>Click <code>Search</code>
<ol>
<li>Displayed will be all BGCs that contain at least one protein associated with the given PFam</li>
</ol>
</li>
</ol>
<h4 id="get-the-accession-ids-for-each-of-proteins-associated-with-an-inputed-pfam">Get the Accession IDs for each of proteins associated with an inputed PFam</h4>
<ol>
<li>Click <code>Replace Input with Output</code></li>
<li>Enter your PFam into the <code>Secondary Input</code> Field</li>
<li>Select <code>Accessions from BGC by PFam</code> as the Output Type</li>
<li>Click <code>Search</code>
<ol>
<li>Displayed will be the Accession ID of each protein that is contains the PFam 'PF10862'</li>
<li>In parentheses is the BGC ID (If one BGC contains multiple proteins with the given PFam they will be displayed on separate lines)</li>
</ol>
</li>
</ol>
<h4 id="get-the-fasta-for-each-of-the-proteins">Get the FASTA for each of the proteins</h4>
<ol>
<li>Click <code>Replace Input with Output</code></li>
<li>Change request type to <code>UniProt</code></li>
<li>Select <code>FASTA</code> as the Output Type</li>
<li>Click <code>Search</code>
<ol>
<li>It may take some time, depending on the number of accessions being searched</li>
<li>Displayed will be the FASTA Sequences for each of the proteins</li>
<li>Some accession ID FASTAs may fail to be obtained due to their absence within UniProt</li>
</ol>
</li>
</ol>
<p>Note: At anyone of these steps the Output can be saved to a text document by clicking <code>Save</code>.</p>
<h3 id="example-2">Example 2:</h3>
<p>You have a GNN of query accessions, you want to compare a target BGC to the other BGCs by comparing PFams.</p>
<h4 id="1-getting-pfams-of-a-bgc">1. Getting PFams of a BGC</h4>
<ol>
<li>Upload the GNN</li>
<li>Select <code>BGC</code> as the <code>Request Type</code></li>
<li>Select <code>BGC-PFam</code> as the <code>Output Type</code></li>
<li>Enter the BGC ID of interest in the input field</li>
<li>Click <code>Search</code>
<ol>
<li>Displayed are all the PFams found within the given BGC</li>
</ol>
</li>
</ol>
<h4 id="2-creating-a-bgc-similarity-table">2. Creating a BGC Similarity Table</h4>
<ol>
<li>Click <code>Replace Input with Output</code></li>
<li>Select <code>BGC-PFam-Similarity</code> as the <code>Output Type</code></li>
<li>Click <code>Search</code>
<ol>
<li>Displayed are the BGC-IDs with their number of PFams they contain from the given list</li>
</ol>
</li>
<li>Click <code>Output Similarity Table</code>
<ol>
<li>You will be prompted to save the *.csv file</li>
<li>The *.csv file provides the searched BGC IDs and whether or not they contain the given PFam ID(s)</li>
</ol>
</li>
</ol>
<p>Note: You may skip the part 1 and simply enter the PFams you wish to compare all BGCs</p>
<p>Note: When searching in part 2, you may enter a PFam into the <code>secondary input</code> to limit the BGCs search to those
that at lease contain that given PFam</p>
<hr>
<h2 id="references">References</h2>
<ol>
<li>Rémi Zallot, Nils Oberg, and John A. Gerlt, The EFI Web Resource for Genomic Enzymology Tools: Leveraging Protein, Genome, and Metagenome Databases to Discover Novel Enzymes and Metabolic Pathways. Biochemistry 2019 58 (41), 4169-4182.</li>
<li>Nils Oberg, Rémi Zallot, and John A. Gerlt, EFI-EST, EFI-GNT, and EFI-CGFP: Enzyme Function Initiative (EFI) Web Resource for Genomic Enzymology Tools. J Mol Biol 2023.</li>
<li>The UniProt Consortium , UniProt: the Universal Protein Knowledgebase in 2023, Nucleic Acids Research, Volume 51, Issue D1, 6 January 2023, Pages D523–D531</li>
</ol>

</body>
</html>
